### PR TITLE
Adding JS Array Examples

### PR DESCRIPTION
**Pull request by walnutwaldo**
The JS Array bindings haven't been fully added yet (exists at https://github.com/Z3Prover/z3/pull/6393) but these examples demonstrate their usage.

Anything I'm forgetting to change/update when adding this example?

### CODE REVIEW

**Comment by walnutwaldo** [comment]
> @microsoft-github-policy-service agree [company="{DeepReason Inc.}"]

**Comment by NikolajBjorner** [comment]
very cool. These will work when z3 gets upgraded to using array bindings in a future release.
I will either merge but comment or wait with the merge.

**Comment by pelikhan** [comment]
it looks like some of the samples don't build. maybe z3 needs update?
image

**Comment by NikolajBjorner** [comment]
correct, I have to deploy a new version of z3 before they have a chance of building

**Review by NikolajBjorner — APPROVED** [review]

**Comment by NikolajBjorner** [comment]
@walnutwaldo - I wonder can you "touch" this pull request to trigger a new build? (I don't see how to do this easily from the UI on this end, so I guess a simple push will do the trick).

**Comment by NikolajBjorner** [comment]
```
Error:  Client bundle compiled with errors therefore further build is impossible.
Error: z3-js runtime error: 3150ee54e4ef0f952c06404754fa1e4130e85798, z3-failed, const arr = Z3.Array.const('arr', Z3.Int.sort(), Z3.Int.sort());
const [idx, val] = Z3.Int.consts('idx val');

const conjecture = arr.store(idx, val).select(idx).eq(val);
Z3.solve(Z3.Not(conjecture));, Error: main.ts (1,16): Property 'Array' does not exist on type 'Context<"main">'.
main.ts (5,10): Argument of type 'Probe<"main">' is not assignable to parameter of type 'Bool<"main">'.
  Type 'Probe<"main">' is missing the following properties from type 'Bool<"main">': not, and, or, xor, and 14 more.
    at runMicrotasks (<anonymous>)
error Command failed with exit code 1.
```

**Comment by NikolajBjorner** [comment]
hmm, we are not on 4.12?

**Comment by NikolajBjorner** [comment]
Right, we are still on 11.0 so the update didn't get pushed.
I have the token.

**Comment by walnutwaldo** [comment]
My bad, the package.json should use the latest z3 now

File: website/package.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
     "remark-math": "3",
     "semantic-release": "^19.0.3",
     "use-editable": "^2.3.3",
-    "z3-solver": "4.11.0"
+    "z3-solver": "4.12.0"
```
Comment (NikolajBjorner) [inline website/package.json:44]: it is 4.12.1 already, but it seems to require an update to yarn.lock as well, which takes a different command.
I guess the upgrade script wasn't fully tested after z3-solver got added to package.json

Comment (NikolajBjorner) [reply website/package.json:44]: the main branch has this issue too

Comment (NikolajBjorner) [reply website/package.json:44]: I will try to resolve this

**Comment by NikolajBjorner** [comment]
OK, this one looks like a legitimate regression:

```
Error: z3-js runtime error: b9be8bd823fa69da4efa6bd43516cef1f3c933c5, z3-failed, const solver = new Z3.Solver();
Error:  Client bundle compiled with errors therefore further build is impossible.
const vector = new Z3.AstVector<Arith>() as AstVector<string, Arith>;
for (let i = 0; i < 5; i++) {
    vector.push(Z3.Int.const(`int__${i}`));
}

const length = vector.length();
for (let i = 0; i < length; i++) {
    solver.add((vector.get(i) as Arith).gt(1));
}
solver.add(Z3.Distinct(...vector));
await solver.check();
solver.model(), Error: main.ts (3,63): Type 'Arith<"main">' does not satisfy the constraint 'Ast<string, unknown>'.
  The types of 'ctx.isModel' are incompatible between these types.
    Type '(obj: unknown) => obj is Model<"main">' is not assignable to type '(obj: unknown) => obj is Model<string>'.
      Type predicate 'obj is Model<"main">' is not assignable to 'obj is Model<string>'.
        Type 'Model<"main">' is not assignable to type 'Model<string>'.
          The types returned by 'entries()' are incompatible between these types.
            Type 'IterableIterator<[number, FuncDecl<"main">]>' is not assignable to type 'IterableIterator<[number, FuncDecl<string>]>'.
              Type '[number, FuncDecl<"main">]' is not assignable to type '[number, FuncDecl<string>]'.
                Type at position 1 in source is not compatible with type at position 1 in target.
                  The types of 'domain(...).cast' are incompatible between these types.
                    Type '(expr: CoercibleToExpr<"main">) => Expr<"main", AnySort<"main">, unknown>' is not assignable to type '(expr: CoercibleToExpr<string>) => Expr<string, AnySort<string>, unknown>'.
                      Types of parameters 'expr' and 'expr' are incompatible.
                        Type 'CoercibleToExpr<string>' is not assignable to type 'CoercibleToExpr<"main">'.
                          Type 'Expr<string, AnySort<string>, unknown>' is not assignable to type 'CoercibleToExpr<"main">'.
                            Type 'Expr<string, AnySort<string>, unknown>' is not assignable to type 'Expr<"main", AnySort<"main">, unknown>'.
                              Types of property 'sort' are incompatible.
                                Type 'AnySort<string>' is not assignable to type 'AnySort<"main">'.
                                  Type 'Sort<string>' is not assignable to type 'AnySort<"main">'.
                                    Type 'Sort<string>' is not assignable to type 'Sort<"main">'.
                                      The types of 'ctx.name' are incompatible between these types.
                                        Type 'string' is not assignable to type '"main"'.
```

**Comment by NikolajBjorner** [comment]
I disabled the failing example from the page (remove tag z3-js) and it builds with 4.12.1.
You should merge your fork and push to it again.

**Comment by walnutwaldo** [comment]
Ah I've actually seen this failing example before. Not sure what causes it but I know how to fix it when I get to my laptop.

**Comment by NikolajBjorner** [comment]
No rush, the build is clear now and ready for testing the pull request.
I presume it exposes a bug in the API bindings. This will trigger a new release, which is relatively easy for me at this point (I introduced a regression today that has to be resolved), 
and then we can get the guide into an updated state.

**Pull request closed without merging**